include/deal.II/grid/connectivity.h

```diff
@@ -0,0 +1,1477 @@
+      {
+        if (structdim == 1)
+          return line_orientation;
+
+        AssertDimension(structdim, 2);
+
+        return quad_orientation;
+      }
+
+      inline const std::vector<unsigned char> &
+      entity_orientations(const unsigned int structdim) const
+      {
+        if (structdim == 1)
+          return line_orientation;
+
+        AssertDimension(structdim, 2);
+
+        return quad_orientation;
+      }
+
+      inline std::vector<unsigned int> &
+      entity_types(const unsigned int structdim)
+      {
+        if (structdim == dim)
+          return cell_types;
+
+        // for vertices/lines the entity types are clear (0/1)
+        AssertDimension(structdim, 2);
+        AssertDimension(dim, 3);
+
+        return quad_types;
+      }
+
+      inline const std::vector<unsigned int> &
+      entity_types(const unsigned int structdim) const
+      {
+        if (structdim == dim)
+          return cell_types;
+
+        // for vertices/lines the entity types are clear (0/1)
+        AssertDimension(structdim, 2);
+        AssertDimension(dim, 3);
+
+        return quad_types;
+      }
+
+      inline CRS<T> &
+      entity_to_entities(const unsigned int from, const unsigned int to)
+      {
+        if (from == dim && to == dim)
+          return neighbors;
+        else if (from == dim && to == dim - 1)
+          return cell_entities;
+        else if (dim == 3 && from == 2 && to == 0)
+          return quad_vertices;
+        else if (dim == 3 && from == 2 && to == 1)
+          return quad_lines;
+        else if (from == 1 && to == 0)
+          return line_vertices;
+
+        Assert(false, ExcNotImplemented());
+
+        return cell_entities;
+      }
+
+      inline const CRS<T> &
+      entity_to_entities(const unsigned int from, const unsigned int to) const
+      {
+        if (from == dim && to == dim)
+          return neighbors;
+        else if (from == dim && to == dim - 1)
+          return cell_entities;
+        else if (dim == 3 && from == 2 && to == 0)
+          return quad_vertices;
+        else if (dim == 3 && from == 2 && to == 1)
+          return quad_lines;
+        else if (from == 1 && to == 0)
+          return line_vertices;
+
+        Assert(false, ExcNotImplemented());
+
+        return cell_entities;
+      }
+
+    private:
+      const unsigned int        dim;
+      std::vector<unsigned int> cell_types;
+
+      CRS<T> line_vertices;
+
+      std::vector<unsigned char> line_orientation;
+
+      CRS<T> quad_vertices;
+      CRS<T> quad_lines;
+
+      std::vector<unsigned char> quad_orientation;
+
+      CRS<T> cell_entities;
+      CRS<T> neighbors;
+
+      std::vector<unsigned int> quad_types;
+    };
+
+
+
+    /**
+     * Create the transposed of a compressed row storage matrix.
+     */
+    template <typename T>
+    CRS<T>
+    transpose(const CRS<T> &con)
+    {
+      auto        con_copy = con;
+      auto &      col      = con_copy.col;
```

Comment:
Wouldn't it be enough to simply copy the `col` array here? I see no reason to copy the `ptr` field.